lib/top/mdx_top.ml

```diff
@@ -397,13 +397,33 @@ let eval t cmd =
           lines := s :: !lines
     in
     let out_phrase' = !Oprint.out_phrase in
+    let print_phrase ppf p =
+      capture ();
+      out_phrase' ppf p;
+      capture ()
+    in
     let out_phrase ppf phr =
       match phr with
       | Outcometree.Ophr_exception _ -> out_phrase' ppf phr
-      | _ ->
-          capture ();
-          out_phrase' ppf phr;
-          capture ()
+      | Outcometree.Ophr_signature signs -> (
+          match suppress with
+          | `Nothing -> print_phrase ppf phr
+          | `Types ->
+              let out_sign = function
+                | (Outcometree.Osig_typext _ | Osig_type _), _ -> ()
+                | _ -> print_phrase ppf phr
+              in
+              List.iter out_sign signs
+          | `Signatures ->
+              let out_sign = function
+                | ( ( Outcometree.Osig_class_type _ | Osig_typext _
+                    | Osig_modtype _ | Osig_type _ ),
+                    _ ) ->
+                    ()
+                | _ -> print_phrase ppf phr
```

Comment:
This causes the double printing. I think this should be done in two steps: filter out items from the tree then call `print_phrase` on the whole result.

Reply:
Good catch, fixed now